app/src/main/java/com/epfl/drawyourpath/userProfile/UserModel.kt

```diff
@@ -0,0 +1,354 @@
+     */
+    fun setNumberOfPathsGoal(nbOfPaths: Int){
+        if(nbOfPaths == 0){
+            throw java.lang.Error("The number of paths goal can't be equal to 0 !")
+        }
+        this.nbOfPathsGoal=nbOfPaths
+    }
+
+    /**
+     * Get the age of the user
+     * @return the age of the user
+     */
+    fun getAge(): Int{
+        var age = LocalDate.now().year-dateOfBirth.year
+        if(LocalDate.now().minusYears(age.toLong())<=dateOfBirth){
+            age -= 1
+        }
+        return age
+    }
+
+    /**
+     * This function will remove the user with username to the friend list
+     * @param username of the user that we want to remove
+     */
+    fun removeFriend(username: String) {
+        if(!friendsList.contains(username)) {
+            throw java.lang.Error("This user is not in the friend list !")
+        }
+        friendsList.remove(username)
+    }
+
+    /**
+     * This function will add the user with username to the friend list. To be added the user must be present in the database.
+     * @param username of the user that we want to add to the friend list
+     */
+    fun addFriend(username: String) {
+        //TODO:this function will be implemented during a next task when the database will be cleaned
+    }
+
+    /**
+     * This function will return the friend list of a user
+     * @return the friend list of the user
+     */
+    fun getFriendList():Map<String, String> {
+        return this.friendsList
+    }
+
+}
+
+/**
+ * Helper function to check the userId and affect it if it's correct
+ * @param userId that correspond to the user
+ * @param database where the userId should be present
+ */
+private fun checkUserId(userId: String, database: Database):Boolean{
+    return database.isUserStoreOnDatabase(userId).get(10, TimeUnit.SECONDS)
+}
+/**
+ * Helper function to check the username and affect it if it's correct
+ * @param userId that correspond to the user
+ * @param username associated to the user profile
+ * @param database where the userId should be present
+ */
+private fun checkUsername(userId: String, username: String, database: Database): Boolean{
+    //TODO: Will be change in my next task when change the database organization
+    return !database.isUserNameAvailable(username).get(10, TimeUnit.SECONDS)
+}
+/**
+ * Helper function to check that the firstname and the surname respect the name format condition of the app
+ * @param name to be checked
+ * @return true if the name is in the correct format, and false otherwise
+ */
+private fun checkName(name: String): Boolean{
+    if(name.find { !it.isLetter() && it != '-' } != null || name.isEmpty()){
+        return false
+    }
+    return true
+}
+
+/**
+ * Helper function to check if the email address is correct
+ * @param email to be checked
+ * @return true is the email is in the correct format, and false otherwise
+ */
+private fun checkEMail(email: String):Boolean {
+    // check for @ char
```

Comment:
Going with our own email address checker is risky. Here, it typically accepts `mail@mail@mail.com`.
The doc shows the existence of `android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()` which might be more robust.

Reply:
I have tested and the function matches doesn't work with this API (I have the same problem with the name format where I have try to used the matcher with REGEX)

Reply:
I've tested 
```kotlin
private fun checkEmail(email: String):Boolean {
    return android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches()
}
```
and it seems to be working as expected.

Reply:
OK, thank you. It's modified.